pkg/ttl/ttlworker/job_manager_integration_test.go

```diff
@@ -1614,3 +1621,225 @@ func TestJobHeartBeatFailNotBlockOthers(t *testing.T) {
+	if s.fault.Load() && s.shouldFault(sql) {
+		return nil, errors.New("fault in test")
+	}
+	return s.Context.GetSQLExecutor().Execute(ctx, sql)
+}
+
+// ExecuteStmt implements sqlexec.SQLExecutor.
+func (s *sessionWithFault) ExecuteStmt(ctx context.Context, stmtNode ast.StmtNode) (sqlexec.RecordSet, error) {
+	// TODO: assert `shouldFault` according to the `stmtNode`. It's not used in TTL so it's fine to keep it now.
+	if s.fault.Load() && s.shouldFault(stmtNode.Text()) {
+		return nil, errors.New("fault in test")
+	}
+	return s.Context.GetSQLExecutor().ExecuteStmt(ctx, stmtNode)
+}
+
+// ExecuteInternal implements sqlexec.SQLExecutor.
+func (s *sessionWithFault) ExecuteInternal(ctx context.Context, sql string, args ...any) (sqlexec.RecordSet, error) {
+	if s.fault.Load() && s.shouldFault(sql) {
+		return nil, errors.New("fault in test")
+	}
+	return s.Context.GetSQLExecutor().ExecuteInternal(ctx, sql, args...)
+}
+
+func (s *sessionWithFault) shouldFault(sql string) bool {
+	// skip some local only sql, ref `getSession()` in `session.go`
+	if strings.HasPrefix(sql, "set tidb_") || strings.HasPrefix(sql, "set @@") {
+		return false
+	}
+
+	return rand.Float64() < s.faultPercent
+}
+
+var _ util.SessionPool = &faultSessionPool{}
+
+type faultSessionPool struct {
+	util.SessionPool
+
+	fault        *atomic.Bool
+	faultPercent float64
+}
+
+func newFaultSessionPool(sp util.SessionPool, percent float64) *faultSessionPool {
+	return &faultSessionPool{
+		SessionPool:  sp,
+		fault:        atomic.NewBool(false),
+		faultPercent: percent,
+	}
+}
+
+// Get implements util.SessionPool.
+func (f *faultSessionPool) Get() (pools.Resource, error) {
+	resource, err := f.SessionPool.Get()
+	if err != nil {
+		return nil, err
+	}
+
+	return &sessionWithFault{
+		Context:      resource.(sessionctx.Context),
+		fault:        f.fault,
+		faultPercent: f.faultPercent,
+	}, nil
+}
+
+// Put implements util.SessionPool.
+func (f *faultSessionPool) Put(se pools.Resource) {
+	f.SessionPool.Put(se.(*sessionWithFault).Context.(pools.Resource))
+}
+
+// SetFault sets whether to enable or disable the fault injection
+func (f *faultSessionPool) SetFault(on bool) {
+	f.fault.Store(on)
+}
+
+func TestGetSessionWithFaultSession(t *testing.T) {
+	_, dom := testkit.CreateMockStoreAndDomain(t)
+	waitAndStopTTLManager(t, dom)
+
+	pool := wrapPoolForTest(dom.SysSessionPool())
+	// defer pool.AssertNoSessionInUse(t)
+	faultPool := newFaultSessionPool(pool, 1)
+
+	se, err := ttlworker.GetSessionForTest(faultPool)
+	require.True(t, se != nil || err != nil)
+
+	faultPool.SetFault(true)
+	se, err = ttlworker.GetSessionForTest(faultPool)
+	require.True(t, se != nil || err != nil)
+}
+
+func TestJobManagerWithFault(t *testing.T) {
+	boostJobScheduleForTest(t)
+
+	store, dom := testkit.CreateMockStoreAndDomain(t)
+	waitAndStopTTLManager(t, dom)
+	tk := testkit.NewTestKit(t, store)
+
+	managerCount := 5
+	testDuration := 10 * time.Second
```

Comment:
TODO: don't run it in CI as it'll need a long time. In my laptop it'll need 500ms for each round. It seems that it's quite unstable (but I still don't know why, investigating) in CI.

Run it nightly, and for at least 10 minutes.

Reply:
We can just add a `t.Skip` here